iproto: document `box.iproto` module (#3766)
tarantool/doc: doc/reference/reference_lua/box_iproto/override.rst

.. _reference_lua-box_iproto_override:

box.iproto.override()
=====================

..  module:: box.iproto

..  function:: override(request_type, handler)

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Set the IPROTO request handler callbacks.

    :param uint32_t     request_type:

>     :param iproto_handler_t     handler:

Reviewer: I guess you copy-pasted this from the C API, while this is not right. The Lua handler has a different interface than the C one, I would describe it separately.

>     :return: 0 on success or -1 on error (check box_error_last())

Reviewer: Ditto. Actually it throws an error.

.. _reference_lua-box_iproto_override:

box.iproto.override()
=====================

..  module:: box.iproto

..  function:: override(request_type, handler)

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Set a new IPROTO request handler callback for the given request type.

    :param number request_type: a request type code. Possible values:

                                *   a type code from :ref:`box.iproto.type <reference_lua-box_iproto_type>` (except

>                                     ``box.iproto.type.UNKNOWN``) -- override the existing request type.

Reviewer: ```suggestion
                                    ``box.iproto.type.UNKNOWN``) -- override the existing request type handler.
```

>                                 *   ``box.iproto.type.UNKNOWN`` -- override an unknown request type.

Reviewer: ```suggestion
                                    ``box.iproto.type.UNKNOWN``) -- override the handler of unknown request types.
```

    :param function handler: IPROTO request handler.
                             The signature of a handler function: ``function(sid, header, body)``, where

                             *  ``sid`` (number): current IPROTO session identifier (see :ref:`box.session.id() <box_session-id>`)
                             *  ``header`` (userdata): a request header encoded as :ref:`msgpack_object <msgpack-object-methods>`
                             *  ``body`` (userdata): a request body encoded as :ref:`msgpack_object <msgpack-object-methods>`

                             Returns ``true`` on success, otherwise ``false``. In case of failure, there is a fallback

>                              to the system handler.

Reviewer: I would mention, that one can indicate an error by throwing an exception. In this sense, retuning `false` means falling back to the system (default) handler, and this is not always a failure, it's just one of the possible options.

..  _reference_lua-box_iproto_override:

box.iproto.override()
=====================

..  module:: box.iproto

..  function:: override(request_type, handler)

    Since version :doc:`2.11.0 </release/2.11.0>`.
    Set a new IPROTO request handler callback for the given request type.

    :param number request_type: a request type code. Possible values:

                                *   a type code from :ref:`box.iproto.type <reference_lua-box_iproto_type>` (except
                                    ``box.iproto.type.UNKNOWN``) -- override the existing request type handler.

                                *   ``box.iproto.type.UNKNOWN`` -- override the handler of unknown request types.

    :param function handler: IPROTO request handler.
                             The signature of a handler function: ``function(sid, header, body)``, where

>                              *  ``sid`` (number): current IPROTO session identifier (see :ref:`box.session.id() <box_session-id>`)

Reviewer 2: ```suggestion
                             *  ``sid`` (number): the current IPROTO session identifier (see :ref:`box.session.id() <box_session-id>`)
```

                             *  ``header`` (userdata): a request header encoded as a :ref:`msgpack_object <msgpack-object-methods>`
                             *  ``body`` (userdata): a request body encoded as a :ref:`msgpack_object <msgpack-object-methods>`

                             Returns ``true`` on success, otherwise ``false``. On ``false``, there is a fallback
                             to the default handler. Also, you can indicate an error by throwing an exception.

>                              In this case, the return value is ``false``, but this does not always means a failure.

Reviewer 2: ```suggestion
                             In this case, the return value is ``false``, but this does not always mean a failure.
```

                             To reset the request handler, set the ``handler`` parameter to ``nil``.

    :return: none

    **Possible errors:**

    If a Lua handler throws an exception, the behavior is similar to that of a remote procedure call.
    The following errors are returned to the client over IPROTO (see `src/lua/utils.h <https://github.com/tarantool/tarantool/blob/dec0e0221e183fa972efa65bb0fb658112f2196f/src/lua/utils.h#L366-L371>`__):

    *   :errcode:`ER_PROC_LUA` -- an exception is thrown from a Lua handler, diagnostic is not set.
    *   diagnostics from ``src/box/errcode.h`` -- an exception is thrown, diagnostic is set.

    For details, see `src/box/errcode.h <https://github.com/tarantool/tarantool/blob/master/src/box/errcode.h>`__.

    **Example:**

    Define a handler function for the ``box.iproto.type.SELECT`` request type:

    ..  code-block:: lua

>         local function iproto_select_handler_lua(sid, header, body)

Reviewer 2: Too many spaces are used for indentation. Here is how the IDE formats this code by default:
<img width="691" alt="image" src="https://github.com/tarantool/doc/assets/38073144/ff8fe8f8-7353-4689-8fb4-67b21b2720bf">
